Test/monadic/MonadicBind.dfy

```diff
@@ -0,0 +1,89 @@
+// RUN: %dafny /compile:3 "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+datatype Option<A> = None | Some(get: A) {
+  function method Bind<B>(f: A --> Option<B>): Option<B>
+    requires this == None || f.requires(this.get)
+    reads if this == None then {} else f.reads(this.get)
+  {
+    if this == None then None else f(get)
+  }
+}
+
+
+
+// -------------------------------
+
+datatype List<A> = Nil | Cons(head: A, tail: List<A>)
+
+function method Head(list: List): Option {
+  if list == Nil then None else Some(list.head)
+}
+function method Tail(list: List): Option<List> {
+  if list == Nil then None else Some(list.tail)
+}
+
+// -------------------------------
+
+class Cell {
+  var data: int
+}
+
+// -------------------------------
+
+// F, G, and H return element 2 of list, if any
+function method F(list: List): Option {
+  match list
+  case Nil => None
+  case Cons(_, cdr) =>
+    match cdr
+    case Nil => None
+    case Cons(_, cddr) =>
+      match cddr
+      case Nil => None
+      case Cons(x, _) => Some(x)
+}
+
+function method G(list: List): Option {
+  Tail(list).Bind(cdr =>
+  Tail(cdr).Bind(cddr =>
+  Head(cddr)))
+}
+
+/* The following uses :- */
+function method H(list: List): Option {
+  var cdr :- Tail(list);
+   var cddr :- Tail(cdr);
+  Head(cddr)
+}
```

Comment:
All the examples involving generics go from a `List<T>` to `List<T>`, where `T` remains the same. I'd also test an example where the function you pass to `Bind` goes from `List<T>` to `List<U>`, where `T` and `U` are different, and also one where it goes from `Foo<T>` to `Bar<U>`, just to see how this interacts with type inference.